In [ ]:
#Descargar libreria openpyxl
! pip install pandas openpyxl

In [ ]:
# Mostrar estaciones cercanas al centro de Madrid
conn = http.client.HTTPSConnection("meteostat.p.rapidapi.com")

headers = {
    'x-rapidapi-key': "154b80484emshe3ce7daf059987fp1aa866jsnc8527d11bc0b",
    'x-rapidapi-host': "meteostat.p.rapidapi.com"
}

conn.request("GET", "/stations/nearby?lat=40.4167&lon=-3.7033", headers=headers) #usando coordinadas encontradas en internet como referencia

res = conn.getresponse()
data = res.read()
print(data.decode("utf-8"))


{"meta":{"generated":"2025-06-24 15:04:07"},"data":[{"id":"08222","name":{"en":"Madrid"},"distance":1735.5},{"id":"08223","name":{"en":"Madrid / Cuatro Vientos"},"distance":7693.8},{"id":"08224","name":{"en":"Madrid / Getafe"},"distance":13033.6},{"id":"08221","name":{"en":"Madrid / Barajas"},"distance":13530.6},{"id":"LEMM0","name":{"en":"Madrid / Barajas de Madrid"},"distance":13766.9},{"id":"08227","name":{"en":"Madrid / Torrejon"},"distance":22712.9},{"id":"08219","name":{"en":"Madri-Colmenar"},"distance":26050.4},{"id":"08215","name":{"en":"Navacerrada"},"distance":48565.9},{"id":"08226","name":{"en":"Guadalajara"},"distance":53216.1},{"id":"08272","name":{"en":"Toledo"},"distance":66218.8}]}


Observamos que aunque contamos con 15 estaciones meteorológicas tomados por el gobierno de España, con meteostat app,  sólo contamos con 4 estaciones:

### Estaciones meteorológicas cercanas a la localización seleccionada

| ID     | Nombre                      | Distancia (m) |
|--------|-----------------------------|----------------|
| 08222  | Madrid                      | 1,735.5        |
| 08223  | Madrid / Cuatro Vientos     | 7,693.8        |
| 08224  | Madrid / Getafe             | 13,033.6       |
| 08221  | Madrid / Barajas            | 13,530.6       |
| LEMM0  | Madrid / Barajas de Madrid  | 13,766.9       |
| 08227  | Madrid / Torrejón           | 22,712.9       |
| 08219  | Madri-Colmenar              | 26,050.4       |
| 08215  | Navacerrada                 | 48,565.9       |
| 08226  | Guadalajara                 | 53,216.1       |
| 08272  | Toledo                      | 66,218.8       |

Vamos ahora a buscar las localizaciones exactas para cada una:


In [ ]:
import http.client
import json

station_ids = [
    "08222", "08223", "08224", "08221", "LEMM0",
    "08227", "08219", "08215", "08226", "08272"
]

headers = {
    "x-rapidapi-key": "",
    "x-rapidapi-host": "meteostat.p.rapidapi.com"
}

conn = http.client.HTTPSConnection("meteostat.p.rapidapi.com")

for sid in station_ids:
    conn.request("GET", f"/stations/meta?id={sid}", headers=headers)
    res = conn.getresponse()
    data = json.loads(res.read().decode("utf-8"))
    station = data.get("data")
    if station:
        name = station.get("name", {}).get("en", "")
        loc = station.get("location", {})
        print(f"{sid} - {name}")
        print(f"  Latitud:  {loc.get('latitude')}")
        print(f"  Longitud: {loc.get('longitude')}")
        print(f"  Altitud:  {loc.get('elevation')} m\n")
    else:
        print(f"{sid} - No hay datos\n")

conn.close()


08222 - Madrid
  Latitud:  40.4167
  Longitud: -3.6833
  Altitud:  655 m

08223 - Madrid / Cuatro Vientos
  Latitud:  40.3833
  Longitud: -3.7833
  Altitud:  690 m

08224 - Madrid / Getafe
  Latitud:  40.3
  Longitud: -3.7167
  Altitud:  620 m

08221 - Madrid / Barajas
  Latitud:  40.45
  Longitud: -3.55
  Altitud:  609 m

LEMM0 - Madrid / Barajas de Madrid
  Latitud:  40.5
  Longitud: -3.5833
  Altitud:  633 m

08227 - Madrid / Torrejon
  Latitud:  40.4833
  Longitud: -3.45
  Altitud:  607 m

08219 - Madri-Colmenar
  Latitud:  40.65
  Longitud: -3.7333
  Altitud:  0 m

08215 - Navacerrada
  Latitud:  40.7833
  Longitud: -4.0167
  Altitud:  1894 m

08226 - Guadalajara
  Latitud:  40.6667
  Longitud: -3.1667
  Altitud:  639 m

08272 - Toledo
  Latitud:  39.8833
  Longitud: -4.05
  Altitud:  515 m



Si tomamos las coordenadas de las estaciones obtenidas por meteostat y las comparamos con las coordenadas de las estaciones de la página del Ministerio para la transición ecológica y el reto demográfico de España, podemos observar cuales estaciones de Meteostat parecen estar más cercanas.

### 🏙️ Estaciones del Ayuntamiento de Madrid con su estación Meteostat más cercana

| COD_LOCAL | LATITUD   | LONGITUD | PROVINCIA | MUNICIPIO | ESTACIÓN | CÓDIGO ESTACIÓN | NOMBRE                                | CÓDIGO METEOSTAT | ESTACIÓN METEOSTAT MÁS CERCANA       |
|-----------|-----------|----------|-----------|-----------|----------|------------------|----------------------------------------|------------------|---------------------------------------|
| 28079004  | 40.42417  | -3.71222 | MADRID    | MADRID    | 4        | ES0115A          | PLAZA DE ESPAÑA                        | 08222           | Madrid                                |
| 28079008  | 40.42167  | -3.68222 | MADRID    | MADRID    | 8        | ES0118A          | ESCUELAS AGUIRRE                       | 08222           | Madrid                                |
| 28079011  | 40.45167  | -3.67722 | MADRID    | MADRID    | 11       | ES0120A          | RAMÓN Y CAJAL                          | 08222           | Madrid                                |
| 28079016  | 40.44     | -3.63917 | MADRID    | MADRID    | 16       | ES0124A          | ARTURO SORIA                           | 08222           | Madrid                                |
| 28079017  | 40.34694  | -3.705   | MADRID    | MADRID    | 17       | ES0125A          | VILLAVERDE                             | 08224           | Madrid / Getafe                       |
| 28079018  | 40.39472  | -3.73194 | MADRID    | MADRID    | 18       | ES0126A          | FAROLILLO                              | 08223           | Madrid / Cuatro Vientos              |
| 28079024  | 40.42     | -3.74917 | MADRID    | MADRID    | 24       | ES1193A          | CASA DE CAMPO                          | 08223           | Madrid / Cuatro Vientos              |
| 28079027  | 40.47472  | -3.575   | MADRID    | MADRID    | 27       | ES1645A          | BARAJAS - PUEBLO                       | LEMM0           | Madrid / Barajas de Madrid           |
| 28079035  | 40.41917  | -3.70333 | MADRID    | MADRID    | 35       | ES1422A          | PLAZA DEL CARMEN                       | 08222           | Madrid                                |
| 28079036  | 40.40806  | -3.64528 | MADRID    | MADRID    | 36       | ES1426A          | MORATALAZ                              | 08222           | Madrid                                |
| 28079038  | 40.44528  | -3.70639 | MADRID    | MADRID    | 38       | ES1525A          | CUATRO CAMINOS-PABLO IGLESIAS         | 08222           | Madrid                                |
| 28079039  | 40.47833  | -3.71139 | MADRID    | MADRID    | 39       | ES1521A          | BARRIO DEL PILAR                       | 08221           | Madrid / Barajas                      |
| 28079040  | 40.38806  | -3.65139 | MADRID    | MADRID    | 40       | ES1532A          | PUENTE DE VALLECAS                     | 08224           | Madrid / Getafe                       |
| 28079044  | 40.49083  | -3.72111 | MADRID    | MADRID    | 44       | ES1893A          | CENTRO CULTURAL ALFREDO KRAUSS        | 08221           | Madrid / Barajas                      |
| 28079045  | 40.39861  | -3.63694 | MADRID    | MADRID    | 45       | ES1894A          | JUNTA MUNICIPAL DE MORATALAZ          | 08222           | Madrid                                |
| 28079047  | 40.39806  | -3.68667 | MADRID    | MADRID    | 47       | ES1937A          | MENDEZ ALVARO                          | 08222           | Madrid                                |
| 28079048  | 40.43972  | -3.69028 | MADRID    | MADRID    | 48       | ES1938A          | CASTELLANA                             | 08222           | Madrid                                |
| 28079049  | 40.40972  | -3.68556 | MADRID    | MADRID    | 49       | ES1939A          | RETIRO                                 | 08222           | Madrid                                |
| 28079050  | 40.46583  | -3.68833 | MADRID    | MADRID    | 50       | ES1940A          | PLAZA CASTILLA-CANAL                   | 08221           | Madrid / Barajas                      |
| 28079054  | 40.37278  | -3.61194 | MADRID    | MADRID    | 54       | ES1941A          | ENSANCHE DE VALLECAS                   | 08224           | Madrid / Getafe                       |
| 28079055  | 40.4625   | -3.58056 | MADRID    | MADRID    | 55       | ES1942A          | URBANIZACION EMBAJADA                 | LEMM0           | Madrid / Barajas de Madrid           |
| 28079056  | 40.3847   | -3.7186  | MADRID    | MADRID    | 56       | ES1943A          | PLAZA ELÍPTICA                         | 08223           | Madrid / Cuatro Vientos              |
| 28079057  | 40.49417  | -3.66028 | MADRID    | MADRID    | 57       | ES1944A          | SANCHINARRO                            | 08221           | Madrid / Barajas                      |
| 28079058  | 40.51806  | -3.77444 | MADRID    | MADRID    | 58       | ES1945A          | EL PARDO                               | 08221           | Madrid / Barajas                      |
| 28079059  | 40.465    | -3.60889 | MADRID    | MADRID    | 59       | ES1946A          | JUAN CARLOS I                          | LEMM0           | Madrid / Barajas de Madrid           |
| 28079060  | 40.50056  | -3.68972 | MADRID    | MADRID    | 60       | ES1947A          | TRES OLIVOS                            | 08221           | Madrid / Barajas                      |
| 28079061  | 40.4025   | -3.695   | MADRID    | MADRID    | 61       | ES1998A          | CENTRO INTEGRADO ARGANZUELA           | 08222           | Madrid                                |

Debido a esto concluimos que las estaciones a utilizar para nuestro análisis en Madrid seran las siguientes:

###  Estaciones Meteostat a utilizar

| Código Meteostat | Nombre                       | Latitud   | Longitud  | Altitud (m) |
|------------------|------------------------------|-----------|-----------|-------------|
| 08222            | Madrid                       | 40.4167   | -3.6833   | 655         |
| 08223            | Madrid / Cuatro Vientos      | 40.3833   | -3.7833   | 690         |
| 08224            | Madrid / Getafe              | 40.3000   | -3.7167   | 620         |
| 08221            | Madrid / Barajas             | 40.4500   | -3.5500   | 609         |
| LEMM0            | Madrid / Barajas de Madrid   | 40.5000   | -3.5833   | 633         |

In [1]:
import http.client, json, pandas as pd
from datetime import datetime, timedelta
import time

# Lista de estaciones meteorológicas de Madrid
stations = ["08222", "08223", "08224", "08221", "LEMM0"]

# Rango de fechas: 1 enero 2022 al 31 diciembre 2023
start_date = datetime(2022, 1, 1)
end_date = datetime(2023, 12, 31)

# Cabeceras de autenticación para RapidAPI
headers = {
    'x-rapidapi-key': "",
    'x-rapidapi-host': "meteostat.p.rapidapi.com"
}

# Devuelve el último día del mes dado
def last_day(date):
    return (date.replace(day=28) + timedelta(days=4)).replace(day=1) - timedelta(days=1)

# Llama a la API con hasta 3 reintentos en caso de error
def fetch_json(url):
    for _ in range(3):
        try:
            conn = http.client.HTTPSConnection("meteostat.p.rapidapi.com")
            conn.request("GET", url, headers=headers)
            res = conn.getresponse()
            data = res.read()
            conn.close()
            return json.loads(data.decode("utf-8"))
        except:
            time.sleep(1)
    return None

# Lista para guardar todos los DataFrames mensuales
all_data = []

# Iterar por estación y por mes
for station in stations:
    current = start_date
    while current <= end_date:
        start = current.replace(day=1)
        end = last_day(start)
        if end > end_date:
            end = end_date

        url = f"/stations/hourly?station={station}&start={start.date()}&end={end.date()}&tz=Europe%2FMadrid"
        data = fetch_json(url)

        if data and "data" in data and data["data"]:
            df = pd.DataFrame(data["data"])
            df["station"] = station
            all_data.append(df)

        current = end + timedelta(days=1)

# Guardar todos los datos tal como vinieron (sin limpieza)
valid_data = [df for df in all_data if df is not None and isinstance(df, pd.DataFrame)]

if valid_data:
    df_all = pd.concat(valid_data, ignore_index=True)
    df_all.to_csv("datos_madrid_2022_2023_meteostat.csv", index=False)

    print("Archivo guardado: datos_meteorologicos_madrid_2022_2023_METEOSTAT.csv")
    print("Total de filas:", len(df_all))
    print("Filas por estación:")
    print(df_all["station"].value_counts())
else:
    print("No se descargaron datos válidos.")


/tmp/ipython-input-1-1845497801.py:62: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  df_all = pd.concat(valid_data, ignore_index=True)


Archivo guardado: datos_meteorologicos_madrid_2022_2023_METEOSTAT.csv
Total de filas: 86184
Filas por estación:
station
08222    17520
08223    17520
08224    17520
08221    17520
LEMM0    16104
Name: count, dtype: int64
